_posts/2022-08-29-sbt-tutorial.md

```diff
@@ -0,0 +1,514 @@
+
+## Advanced configuration in build.sbt
+In the previous section, we saw configurations like _name_, _organization_, etc. The values we set there are applicable for the entire project. SBT also let's us to do some configurations per scope. 
+For example, by default sbt runs all the tests in parallel. If we somehow want to avoid that and run the tests in sequential way, we can set the config option for that. SBT uses the configuration key _parallelExecution_ for that and it is under the scope of Test. 
+
+So, in build.sbt, we can use the setting as:
+```
+Test / parallelExecution := false
+```
+This will set the value to false within the scope of test.
+
+There are many such configurations which are configured using the scope. You can think of this something in the lines of _Test.parallelExecution_.
+
+## Multi Module Project
+So far we have used a single module project. SBT allows to create multi-module project. This way, we can clearly separate different modules, but we can aggregate and combine different modules together to make a big project. Let's create a simple multi-module project first. For that, we can create a build.sbt file as below within a new empty directory.
+
+```
+ThisBuild / scalaVersion := "2.13.8"
+ThisBuild / version := "1.0"
+ThisBuild / name := "multi-module"
+ThisBuild / organization := "com.rockthejvm"
+
+lazy val module_1 = (project in file("module-1"))
+lazy val module_2 = (project in file("module-2"))
+```
+Now, when we can save this file and hit _sbt_ command in the project directoy. This will import the project based on the build.sbt we created and will also create 2 sub directores as _module-1_ and _module-2_ within the directory. The value provided in _file()_ is used to create the sub module name. 
+
+However, as of now there is no relationship between any of the modules. We can explicitly combine both the sub modules together and link to the parent project by adding a new line as belwo to the build.sbt:
+```
+lazy val root = (project in file("."))
+  .aggregate(util, core)
+```
+
+In the same way, we can define dependency between multiple submodules. We can make module-2 to depend on module-1 using dependsOn method:
+
+```
+lazy val module_2 = (project in file("module-2")).dependsOn(module_1)
+```
+
+We can also provide settings for each sub module differently using the settings method:
+
+```
+lazy val module_2 = (project in file("module-2")).settings(
+  libraryDependencies += "com.typesafe" % "config" % "1.4.2"
+)
+```
+
+## Multi-Module Build Best Practises
+We need to first identify and get clarity on different modules. Let's assume that we are building an application that contains database access, HTTP services, utilities etc. 
+Each of these can be separated as a  module. Then we can combine the different parts if one is dependant on another.
+
+Some of the best practices are:
+- single build.sbt file and provide all sub-module information in that sbt file.
```

Comment:
Rephrase of the list to make it "parallel", as now the bullets are disproportionate and don't have the same structure:

> - a single `build.sbt` file with all sub-module information inside
> - common settings at the top (e.g. Scala version, org name, etc)
> - `ThisBuild` to ensure the settings are applied to the entire project including submodules
(cut the code below this bullet, the reader knows this already, because you've introduced them)
> - variables for common library used across multiple sub-modules, applied as module `libraryDependencies`
(put this one first)
> - same for common settings, applied as `settings`
(remove the code block below, I think this is also clear)